blob/blob.go

```diff
@@ -904,9 +909,26 @@ type SignedURLOptions struct {
 	// Expiry sets how long the returned URL is valid for.
 	// Defaults to DefaultSignedURLExpiry.
 	Expiry time.Duration
+
 	// Method is the HTTP method that can be used on the URL; one of "GET", "PUT",
 	// or "DELETE". Defaults to "GET".
 	Method string
+
+	// ContentType specifies the Content-Type HTTP header the user agent is
+	// permitted to use in the PUT request. It must match exactly. See
+	// EnforceAbsentContentType for behavior when ContentType is the empty string.
+	//
+	// Must be empty for non-PUT requests.
+	ContentType string
+
+	// If EnforceAbsentContentType is true and ContentType is the empty string,
+	// then PUTing to the signed URL will fail if the Content-Type header is
+	// present. If EnforceAbsentContentType is false and ContentType is the empty
+	// string, then PUTing without a Content-Type header will succeed, but it is
+	// implementation-specific whether providing a Content-Type header will fail.
```

Comment:
Why is it implementation-specific? Can't we make it so that the behavior is consistent across providers, but sometimes you get Unimplemented?

Reply:
In particular, the backwards-compatible behavior of `("", false)` puts you in the "implementation-specific" bucket, which is unfortunate. It should instead not enforce anything, similar to the current release. 

How about this:

```
// ContentType specifies the Content-Type HTTP header the user agent is
// permitted to use in the PUT request. It must match exactly. See
// EnforceAbsentContentType for behavior when ContentType is the empty string.
// Some providers do not support this and will return an Unimplemented error.
//
// Must be empty for non-PUT requests.

and

// EnforceAbsenceContentType determines behavior when ContentType is the empty string.
//
// If true, PUTing to the signed URL will fail if any Content-Type header is provided.
// Some providers do not support this and will return an Unimplemented error.
//
// If false, no check of the `Content-Type` header is performed.
//
// Must be false for non-PUT requests.
```

IIUC, the Azure implementation matches this already, but the others would need to be updated to look at EnforceAbsentContentType.

Reply:
> In particular, the backwards-compatible behavior of `("", false)` puts you in the "implementation-specific" bucket, which is unfortunate. It should instead not enforce anything, similar to the current release.

I understand the behavior is unfortunate, but that's actually how I encountered this problem in a real-world app. Your understanding is incorrect: GCS and S3 always enforce the Content-Type header. If you use v0.18.0 of CDK and create a signed PUT URL with no Content-Type and then try to upload with a Content-Type header set to something, it will always fail. There is no way AFAICT to disable the behavior, nor do I think it's actually desirable from a security perspective.

The end-user impact of this is really hard to debug, too, since GCS necessarily gives a generic HTTP 403 because it's supposed to be indistinguishable from any other invalid signed URL.

> IIUC, the Azure implementation matches this already, but the others would need to be updated to look at EnforceAbsentContentType.

You'll notice that I don't even check `EnforceAbsentContentType` in GCS and S3 drivers, since there's nothing meaningful they can do with the information.

Reply:
Ah, got it -- I misunderstood the current behavior. This looks OK.